### ISSUE TITLE
SQLiteDatabaseCorruptException: database disk image is malformed (code 11)

### ISSUE THREAD

**Issue by mtrakal** [couchbase/couchbase-lite-android#985]
```
    compile 'com.couchbase.lite:couchbase-lite-android:1.3.1-16'
    compile 'com.couchbase.lite:couchbase-lite-android-forestdb:1.3.1-16'
```
We release new version with 1.3.1-16 (previous was 1.2.1). On Android 5 and 6 we get this error with malformed database (on 7 we use forestDB).

It come from Lollipop and Marshmallow (both use SQLite).
From our code it looks that app is OOM, because database return null object instead of item in.
`StandingOrderGeneratorService.java:227` get records from DB and add them to array. Than it check, that is item null (if null, continue). On this continue it's last part of our code.

I'll try to replicate by myself, this crash comes from users.
```
Fatal Exception: java.lang.RuntimeException: An error occurred while executing doInBackground()
       at android.os.AsyncTask$3.done(AsyncTask.java:309)
       at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
       at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
       at java.util.concurrent.FutureTask.run(FutureTask.java:242)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
       at java.lang.Thread.run(Thread.java:818)
Caused by com.couchbase.lite.internal.database.sqlite.exception.SQLiteDatabaseCorruptException: database disk image is malformed (code 11)
       at com.couchbase.lite.internal.database.sqlite.SQLiteQueryCursor.nativeMoveToNext(SQLiteQueryCursor.java)
       at com.couchbase.lite.internal.database.sqlite.SQLiteQueryCursor.moveToNext(SQLiteQueryCursor.java:58)
       at com.couchbase.lite.storage.SQLiteStorageEngineBase$SQLiteCursor.moveToNext(SQLiteStorageEngineBase.java:257)
       at com.couchbase.lite.store.SQLiteStore.getDocument(SQLiteStore.java:767)
       at com.couchbase.lite.Database.getDocument(Database.java:1454)
       at com.couchbase.lite.Document.getRevision(Document.java:308)
       at com.couchbase.lite.Document.getCurrentRevision(Document.java:205)
       at com.couchbase.lite.Document.getProperties(Document.java:255)
       at com.droid4you.application.wallet.v3.memory.VogelRecord.getRecord(VogelRecord.java:50)
       at com.droid4you.application.wallet.service.StandingOrderGeneratorService$1.onWork(StandingOrderGeneratorService.java:227)
       at com.droid4you.application.wallet.v3.memory.Vogel.run(Vogel.java:244)
       at com.droid4you.application.wallet.v3.memory.Vogel.run(Vogel.java:204)
       at com.droid4you.application.wallet.service.StandingOrderGeneratorService$2.doInBackground(StandingOrderGeneratorService.java:240)
       at com.droid4you.application.wallet.service.StandingOrderGeneratorService$2.doInBackground(StandingOrderGeneratorService.java:236)
       at android.os.AsyncTask$2.call(AsyncTask.java:295)
       at java.util.concurrent.FutureTask.run(FutureTask.java:237)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
       at java.lang.Thread.run(Thread.java:818)
```

**Comment by hideki** [comment]
Note: 

If this issue is view/query related, the issue should be already addressed.
https://github.com/couchbase/couchbase-lite-android/issues/846

If this is non-view/query related issue, there is other scenario that could causes `database disk image is malformed`.

Currently I am not sure how OOM causes `database disk image is malformed` with SQLite.

LINK: http://www.sqlite.org/lockingv3.html#how_to_corrupt

**Comment by pasin** [comment]
@mtrakal does this happen on some specific devices?

**Comment by mtrakal** [comment]
Hi @pasin ,

There is a link on crashes, its possible click on Manufacturer and see most crashed devices http://crashes.to/s/40365dbe08f

With OOM it's just assumption. Because CBL should not return null object in this case. But it return.

Code which is used is like:

```
Cursor c = db.query(SqlRecordMapping.TABLE_RECORDS,null, getFilter(),...);
while (cursor.moveToNext()) {
            Record record = SqlRecordMapping.getRecord(cursor);
            list.add(record);
}
cursor.close();

for (Record record : list) {
                    if (record == null || TextUtils.isEmpty(record.standingOrderId)) {
                        continue;
                    }
...
}
```

VogelRecord.java:50 is: getDatabase().getDocument(record.id).getProperties() // this is last code from us

**Comment by hideki** [comment]
Inputs from @borrrden who works for CBL .NET

.NET also faced `database disk image is malformed` issue before.

```
It was using prepared sqlite statements outside of the protected thread
sqlite3_step is *not* threadsafe unless the connection is compiled with thread safety

The way I solved it is simply by making both connections fully thread safe (serialized mode) 
The tricky thing is, it is not actually corrupt necessarily.  The sqlite3_step is just reading garbage because it is reading from a shared area with not enough thread safety
```

**Comment by hideki** [comment]
From crash log, `SQLiteDatabaseCorruptException` is thrown from `nativeMoveToNext()` method. Like CBL .NET, error occurs in `sqlite3_step(statement)` method.

https://github.com/couchbase/couchbase-lite-java-native/blob/688a26047c26c66cdbac68ed33fe2407eb93cd71/jni/source/com_couchbase_lite_internal_database_sqlite_SQLiteQueryCursor.cpp#L21-L33
```
JNIEXPORT jboolean JNICALL Java_com_couchbase_lite_internal_database_sqlite_SQLiteQueryCursor_nativeMoveToNext
(JNIEnv* env, jclass clazz, jlong statementPtr) {
    sqlite3_stmt* statement = reinterpret_cast<sqlite3_stmt*>(statementPtr);
    int err = sqlite3_step(statement);
    if (err == SQLITE_ROW) {
        return true;
    } else if (err == SQLITE_DONE) {
        return false;
    } else {
        throw_sqlite3_exception(env, sqlite3_db_handle(statement), NULL);
    }
    return false;
}
```

**Comment by hideki** [comment]
CBL .NET faced same issue. CBL .NET addressed this issue by using `SQLITE_OPEN_FULLMUTEX ` (see https://www.sqlite.org/threadsafe.html) which is same thing with compiling SQLite with Serialized mode (-DSQLITE_THREADSAFE=1 or no parameter).

https://github.com/couchbase/couchbase-lite-net/commit/f8ae2121012ea03fac67932b60a7fc34c45edc40

**Comment by pasin** [comment]
Currently we are building SQLite with Multi-thread mode which is the same mode as the one comes with Android SDK.

'''
Multi-thread. In this mode, SQLite can be safely used by multiple threads provided that no single database connection is used simultaneously in two or more threads.
'''

We copied Android SQLiteDatabase wrapper that provides local thread bound connection pool to prevent a connection to be used by multiple thread.

If the root cause is due to threading issues, it could possibly mean that:

1. There could be a bug in SQLiteDatabase connection pool or related modules.

2. There is a bug in CBL that violates the usage of SQLiteDatabse. For example, Does CBL allow 2 database objects of the same databse name got created (from REST or Native API)? Or Does CBL create a prepare statement on one thread but uses by anither thread?

**Comment by hideki** [comment]
Multithreaded vs. serialized
http://sqlite.1065341.n5.nabble.com/Multithreaded-vs-serialized-td46575.html

> In serialized mode, each database handle (sqlite3*) has an internal 
mutex. Each time any API call is made on that database handle, the 
mutex is obtained, the work of the API call is done, and the mutex 
is released.

`Serialized mode has a mutex in each database handle (sqlite3*).`  It makes each connection thread safe. Each connection should not be influenced each other by `Serialized` mode.

> In multi-thread mode, this database handle mutex is omitted. So things 
run slightly faster because there is one less mutex grab each time 
an API call is made.

Acquiring mutex every request should not make SQL operation significantly slow.

**Comment by pasin** [comment]
@hideki do we have the code similar to .NET that uses a cursor on multiple theads? I'm wondering why CBL .NET does't fix that but instead turning to open the DB with serialize mode (ATT: @borrrden).

**Comment by hideki** [comment]
From code review point of view, CBL java does not allow multiple threads access to same connection or sharing cursor between threads.

I don't know details of CBL .NET implementation.

**Comment by hideki** [comment]
TODOs:

Code Review:
1. Review Cursor implementation if it has bug. ex: Cursor could return connection before closed.
1. Review CBL java core codes if any cursor is shared by multiple threads.

Write Unit Test:
1. Port test .NET unit test which tests multi-threads execute view/query operation. https://github.com/couchbase/couchbase-lite-net/blob/master/src/Couchbase.Lite.Tests.Shared/ViewsTest.cs#L399

**Comment by hideki** [comment]
Hi @mtrakal 

Opening the database with `Serialized` mode (see https://www.sqlite.org/threadsafe.html) could solve this issue. If you could build Couchbase Lite from source code, could you please try this solution? Needs to add `SQLITE_OPEN_FULLMUTEX` option to `sqliteFlags` before calling `sqlite3_open_v2()`?

```
sqliteFlags |= SQLITE_OPEN_FULLMUTEX;
```
https://github.com/couchbase/couchbase-lite-java-native/blob/688a26047c26c66cdbac68ed33fe2407eb93cd71/jni/source/com_couchbase_lite_internal_database_sqlite_SQLiteConnection.cpp#L90

**Comment by mtrakal** [comment]
Hi @hideki ,

I'm able to build CBL by myself, but I'm not able to reproduce this issue. It comes from users, so I have no idea how to test it (expect release new version of app to public).

**Comment by hideki** [comment]
Hi @mtrakal,

I understand your concerns. So far, we also can not reproduce the issue internally.

From the Fabric data, 99% of all `SQLiteDatabaseCorruptException` occurs with Android 5 or 6. I believe the issue is caused by allowing multiple SQLite connections for Android 5 and 6. Till 1.2.1, maximum connection is always 1 for all Android versions.

If your application does not causes large data pull replication or large data indexing, you could change maximum connection to 1 from 4 by modifying following line of code.  
https://github.com/couchbase/couchbase-lite-android/blob/master/src/main/java/com/couchbase/lite/android/AndroidSQLiteStorageEngine.java#L58
If your application causes large data insertion or update, SQLite read operation from other thread will be blocked by the write operation.

Thanks,
Hideki

**Comment by hideki** [comment]
Note to Self:

From code review, 
- `SQLiteConnection ` is not thread-safe. It holds `mIsPrimaryConnection` variable to check if connection is primary connection (readonly or read/write mode).
- `SQLiteConnectionPool` is thread-safe, and it guarantees it does not return same connection to multiple threads. 
- `SQLiteSession` instance is per a thread. This is guaranteed by ThreadLocal. It also guarantees single connection at once by reusing connection if it already owns connection.
```java
    private void acquireConnection(String sql, int connectionFlags,
            CancellationSignal cancellationSignal) {
        if (mConnection == null) {
            assert mConnectionUseCount == 0;
            mConnection = mConnectionPool.acquireConnection(sql, connectionFlags,
                    cancellationSignal); // might throw
            mConnectionFlags = connectionFlags;
        }
        mConnectionUseCount += 1;
    }
```
https://github.com/couchbase/couchbase-lite-java-core/blob/4cc2f7d66a9be58ca774d102497b7a725885e168/vendor/sqlite/src/java/com/couchbase/lite/internal/database/sqlite/SQLiteSession.java#L845-L854

It means that `SQLiteSession` could re-use `SQLiteConnection` for read operation (SELECT) as  `SQLiteConnection` for write operation (INSERT/UPDATE/DELETE/...). For example, Iterating records by `SELECT` query, internally calling `INSERT/UPDATE/DELETE`.  `SQLiteViewStore .updateIndexes(...)` has this type of operation.
https://github.com/couchbase/couchbase-lite-java-core/blob/4cc2f7d66a9be58ca774d102497b7a725885e168/src/main/java/com/couchbase/lite/store/SQLiteViewStore.java#L248-L582

This codes (updateIndexes) might not cause the problem if the application is single thread or application is running with single SQLiteConnection instance mode. But, if multiple threads call `updateIndexes(...)` method concurrently, multiple write operations could be executed to the database. 
Currently we don't know if multiple write operations against same database/same table work fine.
I believe SQLite write operation should be synchronized, and it potentially could fix this issue.

@pasin, do you have any comment?

**Comment by mtrakal** [comment]
We found few users which application data consume all internal storage capacity (cbl).
Our application stuck in infinite loop which resolve conflicts:

`changeEvent.getChanges()` contains one document
after call `resolveConflictsIfExists > checkAndResolveConflict`
we receive back in `changed` TWO same documents (both has same documentId, winningRevisionId and both are isConflict = true) and some really old (maybe deleted document)
So next iterateion make 4 documents, next 8 documents, ... growing to infinite, instead of removing the one conflicted document from list `changeEvent.getChanges()`

We call:
```
private Database.ChangeListener getChangeListener() {
        return new Database.ChangeListener() {
            @Override
            public void changed(Database.ChangeEvent changeEvent) {
                for (DocumentChange documentChange : changeEvent.getChanges()) {
                    Map<String, Object> properties = getDatabase().getDocument(documentChange.getDocumentId()).getProperties();
                    resolveConflictsIfExists(properties, documentChange);
...
```

```
private void resolveConflictsIfExists(Map<String, Object> properties, DocumentChange documentChange) {
        ...
            if (documentChange.isConflict()) {
                Document document = getDatabase().getDocument(documentChange.getDocumentId());
                checkAndResolveConflict(document);
            }
        }
    }
```

We resolve conflicts as use current revision (update them and save).
```
public static void checkAndResolveConflict(Document document) {
        try {
            for (SavedRevision conflictingRevision : document.getConflictingRevisions()) {
                UnsavedRevision newRevision = conflictingRevision.createRevision();
                if (conflictingRevision.equals(document.getCurrentRevision())) {
                    YablohnLogger.d(TAG, "Resolving conflict");
                    // time to merge conflicts
                } else {
                    newRevision.setIsDeletion(true);
                }
                newRevision.save(true);
            }
        } catch (CouchbaseLiteException e) {
            YablohnLogger.e(TAG, "error while resolving conflicts", e);
        }
    }
```

We reduce a lot of duplication by: 
```
if(!documentChange.isCurrentRevision()) {
                        continue;
                    }
                    if(documentChange.getAddedRevision() != null && documentChange.getAddedRevision().isDeleted()) {
                        continue;
                    }
```
But still receive the items as on screen. One looks deleted with revision 4. One with every call checkAndResolveConflict increase revision number (but still have `isConflict = true`

![clipboard02](https://cloud.githubusercontent.com/assets/1458891/18208433/c68e32d6-712f-11e6-9e2c-50aecd4a6fc6.png)

After all storage is filled CBL can not create a new revision and force SQLiteDatabaseCorruptException

**Comment by hideki** [comment]
Note to Self:

According to SQLite, hard disk issue has higher chance to make corrupt database. `Out of disk space` could cause this `malformed disk image` issue. 
- https://www.sqlite.org/atomiccommit.html#sect_9_0
- https://www.sqlite.org/howtocorrupt.html


ToDo:
- Reproduce infinite calling each other by  `ChangeListener` callback and Document update.

**Comment by hideki** [comment]
Hi @mtrakal,

I have tried to reproduce the issue of `isConflict=true after setting DELETE true`. I tried following three scenarios:
- Create conflict in client programmatically, and solve it in `ChangeListener.changed()`. 
- Create conflict by client and Sync Gateway, and solve it in `ChangeListener.changed()`. 
- Create conflict by client and CouchDB, and solve it in `ChangeListener.changed()`.

So far, I could not reproduce the problem. So I need additional inputs to reproduce the problem from you.

From the image you posted, 
- Why are `isConflict` variables `true`. It should be `false` after solving conflict.
- Why do both revisions have same `sequence` number (`262`). `sequence` number should be unique per revisions/doc.
- `source` variables are `null`. So I assume conflict occurs in the client app. Not with replication. <- is this correct??

`sequence` number is actually row id of `revs` table of SQLite. As both revisions have same sequence number, it indicates that the database file was already corrupted when you captured image?

Thanks,
Hideki

**Comment by hideki** [comment]
Hi @mtrakal,

Since v1.2.0, CBL Android/Java ported SQLiteDatabase of Android (https://github.com/android/platform_frameworks_base/tree/master/core/java/android/database) into CBL Android/Java. But CBL makes disable `DefaultDatabaseErrorHandler`.
- https://github.com/android/platform_frameworks_base/blob/e71ecb2c4df15f727f51a0e1b65459f071853e35/core/java/android/database/sqlite/SQLiteDatabase.java#L259
- https://github.com/couchbase/couchbase-lite-java-core/blob/f85b64f5d65b6a6d83d5de0d1763a062e24590e1/vendor/sqlite/src/java/com/couchbase/lite/internal/database/sqlite/SQLiteDatabase.java#L253

Original implementation of `SQLiteDatabase/DefaultDatabaseErrorHandler` deletes the corrupted database if database is corrupted. 
Till CBL 1.1.x, CBL Android/Java uses SQLiteDatabase that Android OS provided. So end-users might encounter database corruption issue, but SQLiteDatabase automatically deleted it. And the app could recover from it.
From CBL 1.2.x, CBL ported SQLiteDatabase, but disabled the corrupted database auto-delete feature, but CBL does not enable multiple SQLite connections, So end-user less frequently hit this issue.

Suggestion: Your app could use `DefaultDatabaseErrorHandler` or implement own `DatabaseErrorHandler` for your app to recover from this. 

We will track how to handle Database error by https://github.com/couchbase/couchbase-lite-java-core/issues/1427

Thanks,
Hideki

**Comment by snej** [comment]
> Out of disk space could cause this malformed disk image issue.

I don't believe that's true. SQLite correctly handles ordinary write errors like out-of-space. In the SQLite docs you linked to, there is no mention of running out of disk space as a cause of corruption.

**Comment by pasin** [comment]
> Why are isConflict variables true. It should be false after solving conflict.

I could (re)produce the case that the DocumentChange of one of the deleted conflicted revisions  is still shown as conflicted when using **ForestDB** but not SQLite (See https://github.com/couchbase/couchbase-lite-java-core/issues/1429).

@mtrakal from the screenshot you provide above, can you confirm the storage engine?

**Comment by mtrakal** [comment]
Hi @hideki ,

for your questions:

I connect to backend couchDB which was corrupted (must have conflict) to new device. It create local db and problem start after first close app (we start resolving conflicts before closing app). I was in offline mode, so replication is not problem.
I think, that's caused by some non-thread safe calling which make two objects with same revision?

- Why are isConflict variables true. It should be false after solving conflict.
I have no idea, it comes from CBL in `public void changed(Database.ChangeEvent changeEvent) {`. In previous version of CBL 1.2.1 we received `isConflict = false` after resolve conflicts. But not with this userspace on 1.3.1.

- Why do both revisions have same sequence number (262). sequence number should be unique per revisions/doc.
Have no idea (and have no idea why is there so old revision with number 4 (maybe original revision before infinite loop)).

- source variables are null. So I assume conflict occurs in the client app. Not with replication. <- is this correct??
Source is really current device, because until closing app (start resolving conflicts) it doesn't fill storage.

I think, that could be relevant full storage and corrupted database, but I'm not sure, just assumption.

@pasin 
My fault, yes, storage in this case was ForestDB (I tested it on Nexus 5x with latest android N, not on Android 5/6 :/) Sorry, maybe different issue? But looks as mostly same behaviour as on Android 5/6 (sqlite). Resolving conflict doesn't work as expected for us.

**Comment by hideki** [comment]
Note: ForestDB isConflict variable issue is tracked by https://github.com/couchbase/couchbase-lite-java-core/issues/1429

**Comment by hideki** [comment]
Hi @mtrakal,

I have some questions. 

I am wondering if multiple SQLite database instances are created, and it could cause database corruption.
- How does the app create a instance of Manager? Is Manager instance one instance per app?
- How does the app get/open database from Manager? Is it operated from threads (AsyncTask)?

Questions from your previous post:

> I connect to backend couchDB which was corrupted (must have conflict) to new device. It create local db and problem start after first close app (we start resolving conflicts before closing app). I was in offline mode, so replication is not problem.

- Can you reproduce the issue with SQLite storage type by yourself now?
- If yes, Does the app not cause DatabaseCorruption exception?
- What does the application do for `resolving conflict`? Is this triggered by `closing app`.

Thanks,
Hideki

**Comment by hideki** [comment]
Note to self:

In case SQLite connection for non-primary (read-only) is used for write operation, SQLiteDatabase writes WARNING message to `logcat`. But it does not throw `SQLiteException` if the database is not read-only database.

Tested with commenting out beginTransaction() and endTransaction() in `SQLiteViewStore.updateIndexes(...)`.

Printed warning message: 
```
09-08 02:11:56.501 3865-3884/com.couchbase.lite.test W/database: SQLiteConnection: The connection is acquired for a read-only statement. This ongoing statement (INSERT INTO 'maps_1' (sequence, key, value) VALUES(?,?,?)) might modify the database with the connection.
09-08 02:11:56.505 3865-3884/com.couchbase.lite.test W/database: SQLiteConnection: The connection is acquired for a read-only statement. This ongoing statement (INSERT INTO 'maps_1' (sequence, key, value) VALUES(?,?,?)) might modify the database with the connection.
09-08 02:11:56.505 3865-3884/com.couchbase.lite.test W/database: SQLiteConnection: The connection is acquired for a read-only statement. This ongoing statement (INSERT INTO 'maps_1' (sequence, key, value) VALUES(?,?,?)) might modify the database with the connection.

```

SQLiteDatabase Code could write warning message.
```java
    private void warnOrthrowIfStatementForbidden(PreparedStatement statement) {
        if (mOnlyAllowReadOnlyOperations && !statement.mReadOnly) {
            if (mConfiguration.isOpenReadOnly()) {
                throw new com.couchbase.lite.internal.database.sqlite.exception.SQLiteException("Cannot execute this statement (" + statement.mSql + ") "
                        + "because it might modify the database but the connection is read-only.");
            } else {
                DLog.w(TAG, "The connection is acquired for a read-only statement. "
                        + "This ongoing statement (" + statement.mSql + ") might modify "
                        + "the database with the connection.");
            }
        }
    }
```
https://github.com/couchbase/couchbase-lite-java-core/blob/f25f9a2cbd2ed4f1bb71e07022e6aaa2643267c2/vendor/sqlite/src/java/com/couchbase/lite/internal/database/sqlite/SQLiteConnection.java#L830-L841

Next Step: 
- Figure out if using non-primary SQLiteConnection for write operation could cause corrupted database.

**Comment by mtrakal** [comment]
> I am wondering if multiple SQLite database instances are created, and it could cause database corruption.

We create an instance in Application context and database can have more instances (but we switch database only in service for initial replication - when user logout from app and login as another user. It's not so common action)

> How does the app create a instance of Manager? Is Manager instance one instance per app?

We use only one Manager which is created from Application.

```
private void init(IReplicable ownerProvider) {
        try {
            mManager = new Manager(new AndroidContext(mContext), Manager.DEFAULT_OPTIONS);
        } catch (IOException e) {
            return;
        }
        connectDatabase(ownerProvider);
    }
```

```
 private void connectDatabase(IReplicable ownerProvider) {
        // create a name for the mDatabase and make sure the name is legal
        if (!Manager.isValidDatabaseName(getLocalDbName(ownerProvider))) {
            return;
        }
        try {
            // same database is already opened
            if (mDatabase != null) {
                if (mDatabase.getName().equals(getLocalDbName(ownerProvider))) {
                    return;
                } else {
                    mPreviousDatabase = mDatabase;
                }
            }
            DatabaseOptions options = new DatabaseOptions();
            options.setCreate(true);
            if ("N".equals(Build.VERSION.CODENAME) || Build.VERSION.SDK_INT > 23) {
                options.setStorageType(Manager.FORESTDB_STORAGE);
            } else {
                options.setStorageType(Manager.SQLITE_STORAGE);
            }
            mDatabase = mManager.openDatabase(getLocalDbName(ownerProvider), options);
        } catch (CouchbaseLiteException e) {
            Log.e(TAG, "Cannot create database", e);
            return;
        }
    }
```

> How does the app get/open database from Manager? Is it operated from threads (AsyncTask)?

We open it in `onCreate` in `Application extends MultiDexApplication`. When we create initial replication we switch db in service.

> I connect to backend couchDB which was corrupted (must have conflict) to new device. It create local db and problem start after first close app (we start resolving conflicts before closing app). I was in offline mode, so replication is not problem.
> Can you reproduce the issue with SQLite storage type by yourself now?
> If yes, Does the app not cause DatabaseCorruption exception?
> What does the application do for resolving conflict? Is this triggered by closing app.

I'll try to reproduce (hope what we have still corrupted database).

Complete resolving is described in: https://github.com/couchbase/couchbase-lite-android/issues/985#issuecomment-244402035
Yes, it's triggered by closing app, it's just trigger to start action.

**Comment by mtrakal** [comment]
I'm sorry, I'm able to reproduce this only on Android N with ForestDB, on Android 5/6 I'm not able to reproduce for account on which we can test it :(.

**Comment by mtrakal** [comment]
```
compile 'com.couchbase.lite:couchbase-lite-android:1.3.1-16'
compile 'com.couchbase.lite:couchbase-lite-android-forestdb:1.3.1-16'
```
Android N, when I flood complete storage by revisions, it stuck on:
```
mDatabase = mManager.openDatabase(getLocalDbName(ownerProvider), options);
Manager:335 > db.open(options);
Database:1251 > store.open();
ForestDbStore:118 > this.forest = new Database(this.forestPath, flags, enAlgorithm, enKey);
Database.class:22 > this._handle = this._open(path, flags, encryptionAlgorithm, encryptionKey);
Database.class:69 > private native long _open(String var1, int var2, int var3, byte[] var4) throws ForestException;
```

There is complete threads report:
[threads_report.txt](https://github.com/couchbase/couchbase-lite-android/files/461481/threads_report.txt)
Our main thread stuck in this code. No error, no exception, just freeze.

**Comment by pasin** [comment]
> Figure out if using non-primary SQLiteConnection for write operation or multiple write SQLite connections could cause corrupted database.

In theory, under WAL mode, it allows concurrent write and read BUT I think it could end up with unexpected data in the database: 

```
WAL mode permits simultaneous readers and writers. It can do this because changes do not overwrite the original database file, but rather go into the separate write-ahead log file. That means that readers can continue to read the old, original, unaltered content from the original database file at the same time that the writer is appending to the write-ahead log. In WAL mode, SQLite exhibits "snapshot isolation". When a read transaction starts, that reader continues to see an unchanging "snapshot" of the database file as it existed at the moment in time when the read transaction started. Any write transactions that commit while the read transaction is active are still invisible to the read transaction, because the reader is seeing a snapshot of database file from a prior moment in time.
```

> However, if multiple threads concurrently write to database, SQLiteDatabase throws SQLiteDatabaseLockedException as database file is locked.

I think this gives us a clue for another issue. We should make sure that we fix all of these scenarios. Maybe instead of warning, making it throws the exception. in CBL iOS, we did an assertion for multilthreading violation.

**Comment by snej** [comment]
> Figure out if using non-primary SQLiteConnection for write operation or multiple write SQLite connections could cause corrupted database

That should not be possible. SQLite's file locking prevents more than one connection from writing ot the file at a time.

**Comment by hideki** [comment]
@mtrakal Thank you for your inputs. As Manager instance is created from Application, multiple Manager instances should not be happen, and multiple Database instances for same database should not be happen.

In this ticket, two issues are reported. One is `SQLiteDatabaseCorruptException`, another one is `isConflicts` variable is still `true` after deleting losing revision. We'd like to treat two issue separately. This ticket is for SQLite database issue. And https://github.com/couchbase/couchbase-lite-java-core/issues/1429 for ForestDB issue. ForestDB issue is reproducible.

**Comment by hideki** [comment]
Note to self:

Experimented that `throwing Exception if a read connection is used for a write operation` with unit tests. I can not observe `Exception` with unit test. So CBL Android/Java code does not do SQLite write operation (INSERT/UPDATE/DELETE) in the read operation (SELECT).

**Comment by hideki** [comment]
Hi @mtrakal,

In [your previous comment](https://github.com/couchbase/couchbase-lite-android/issues/985#issuecomment-243688020), you shared code snippet of `StandingOrderGeneratorService.java`. In this snippet, it indicates the application uses SQLite beside the Couchbase Lite. Could you please explain how SQLite database and Couchbase Lite co-exist? I am wondering how `db` is instantiated? Is its class crated from  `android.database.sqlite.SQLiteDatabase`  or `com.couchbase.lite.internal.database.sqlite.SQLiteDatabase`. Is SQLite database file is shared by SQLite and Couchbase Lite? or is it different SQLite database file?
```java
Cursor c = db.query(SqlRecordMapping.TABLE_RECORDS,null, getFilter(),...);
while (cursor.moveToNext()) {
            Record record = SqlRecordMapping.getRecord(cursor);
            list.add(record);
}
cursor.close();
```

Thanks,
Hideki

**Comment by hideki** [comment]
@mtrakal,

Another potential scenario might cause `SQLiteDatabaseCorruptException`

Some of end users, who uses CBL v1.x, needed to migrate CBL database from v1.x to v1.2/v1.3. It is processed by `DatabaseUpgrade` class. If user's data is large, it takes a while. Sometimes user kills the application. Until  fixing the issue #888, CBL kept using the partially updated newly created database. It caused issue of data loss (#888). Possibly, it might cause Database corruption  too because the application is killed middle of data insertion by `DatabaseUpgrade`.

According to @mtrakal, the application integrates CBL android `1.3.1-16`, which was built from commit (ee10077) (see: http://mobile.jenkins.couchbase.com/view/Couchbase_Lite/job/couchbase-lite-android-master/16/). `1.3.1-16` does not include the fix of #888.

If killing the application during database upgrade is a root cause of database corruption, using latest build of CBL Android (1.3.1-30) could fix this issue.

**Comment by snej** [comment]
No, killing the process cannot corrupt a SQLite database.

**Comment by hideki** [comment]
Following is `SQLiteDatabaseCorruptException` stack trace which is commented for other ticket. `SQLiteDatabaseCorruptException` is thrown while reading data from `revs` table that was created with CBL v1.1.0. In `DatabaseUpgrade.importData()` method, reading data from `docs` table is processed before reading data from `revs` table. But the exception was not thrown while reading data from `docs` table. Is database corruption detected per table? or Could database corruption error occur without actual database file corruption? @borrrden shared his experience he observed SQLite library return database corruption error without actual database file corruption.

https://github.com/couchbase/couchbase-lite-android/issues/846#issuecomment-241761170
```
com.couchbase.lite.internal.database.sqlite.exception.SQLiteDatabaseCorruptException: database disk image is malformed (code 11)
  at com.couchbase.lite.internal.database.sqlite.SQLiteQueryCursor.nativeMoveToNext(Native Method)
  at com.couchbase.lite.internal.database.sqlite.SQLiteQueryCursor.moveToNext(SQLiteQueryCursor.java:58)
  at com.couchbase.lite.storage.SQLiteStorageEngineBase$SQLiteCursor.moveToNext(SQLiteStorageEngineBase.java:257)
  at com.couchbase.lite.DatabaseUpgrade.importDoc(DatabaseUpgrade.java:223)
  at com.couchbase.lite.DatabaseUpgrade.access$000(DatabaseUpgrade.java:38)
  at com.couchbase.lite.DatabaseUpgrade$1.run(DatabaseUpgrade.java:118)
  at com.couchbase.lite.store.SQLiteStore.runInTransaction(SQLiteStore.java:717)
  at com.couchbase.lite.Database.runInTransaction(Database.java:565)
  at com.couchbase.lite.DatabaseUpgrade.importData(DatabaseUpgrade.java:111)
  at com.couchbase.lite.Manager.upgradeDatabase(Manager.java:839)
  at com.couchbase.lite.Manager.upgradeOldDatabaseFiles(Manager.java:817)
  at com.couchbase.lite.Manager.<init>(Manager.java:130)
```

**Comment by pasin** [comment]
@hideki I have seen this : https://chromium.googlesource.com/android_tools/+/18728e9dd/sdk/sources/android-19/android/database/DatabaseErrorHandlerTest.java#100.

I couldn't find the same source from Android SDK, but I might tell something.

**Comment by hideki** [comment]
@pasin 
https://github.com/android/platform_frameworks_base/blob/master/core/java/android/database/sqlite/SQLiteDatabase.java#L2141-L2176
```java
    public boolean isDatabaseIntegrityOk() {
        acquireReference();
        try {
            List<Pair<String, String>> attachedDbs = null;
            try {
                attachedDbs = getAttachedDbs();
                if (attachedDbs == null) {
                    throw new IllegalStateException("databaselist for: " + getPath() + " couldn't " +
                            "be retrieved. probably because the database is closed");
                }
            } catch (SQLiteException e) {
                // can't get attachedDb list. do integrity check on the main database
                attachedDbs = new ArrayList<Pair<String, String>>();
                attachedDbs.add(new Pair<String, String>("main", getPath()));
            }

            for (int i = 0; i < attachedDbs.size(); i++) {
                Pair<String, String> p = attachedDbs.get(i);
                SQLiteStatement prog = null;
                try {
                    prog = compileStatement("PRAGMA " + p.first + ".integrity_check(1);");
                    String rslt = prog.simpleQueryForString();
                    if (!rslt.equalsIgnoreCase("ok")) {
                        // integrity_checker failed on main or attached databases
                        Log.e(TAG, "PRAGMA integrity_check on " + p.second + " returned: " + rslt);
                        return false;
                    }
                } finally {
                    if (prog != null) prog.close();
                }
            }
        } finally {
            releaseReference();
        }
        return true;
    }
```

**Comment by hideki** [comment]
@pasin, thank you for finding the test case.
With current information, I am sure that SQLite native library returns `SQLITE_CORRUPT(11)`, but I am not sure if the database file is actually corrupted, as non of reported `SQLiteDatabaseCorruptException` exception is thrown from write operations (INSERT/DELETE/UPDATE/...). Maybe it is good idea to implement `DatabaseErrorHandler` to check the return value from `SQLiteDatabase. isDatabaseIntegrityOk()` method to identify if the database is really corrupted when `SQLiteDatabaseCorruptException` is thrown.

**Comment by hideki** [comment]
@mtrakal,

Following links indicates that `Multiple copies of SQLite linked into the same application` could cause database corruption. We really would like to understand how your app uses Couchbase Lite and Android SQLite. From CBL v1.3.0, CBL Android includes SQLite library to support Android N that does not allow application to access Android provided native libraries.
 
https://discuss.zetetic.net/t/is-the-multiple-sqlite-problem-an-issue-for-sqlcipher-for-android/1449/2
http://ericsink.com/entries/sqlite_android_n.html
https://www.sqlite.org/howtocorrupt.html
http://ericsink.com/entries/multiple_sqlite_problem.html

**Comment by pasin** [comment]
From https://github.com/couchbase/couchbase-lite-android/issues/966#issuecomment-240658168 that @mtrakal have posted, it seems like some of revision ids could contains non-ascii.  Can this cause an issue with the RevID collator and result to the database corruption?

**Comment by borrrden** [comment]
> Is database corruption detected per table?

Not even per table, it is detected at the moment it tries to read from a certain area and does not get something formatted the way it expects.

**Comment by mtrakal** [comment]
Hi @hideki 

> In your previous comment, you shared code snippet of StandingOrderGeneratorService.java. In this snippet, it indicates the application uses SQLite beside the Couchbase Lite. Could you please explain how SQLite database and Couchbase Lite co-exist? I am wondering how db is instantiated? Is its instantiated crated from android.database.sqlite.SQLiteDatabase class or com.couchbase.lite.internal.database.sqlite.SQLiteDatabase class? Is SQLite database file shared by SQLite and Couchbase Lite? Or do SQLite and Couchbase Lite have own database file?

We use sqlite `android.database.sqlite.SQLiteDatabase` for faster querying, we duplicate just some most required columns from CBL documents to our (another than CBL internal sqlite) database for advanced queries.
We have CBL for persisting data, when some update come to CBL we update our sqlite.

**Comment by hideki** [comment]
Hi @mtrakal, 
Thank you for your comment. I need to confirm one more thing. Does `android.database.sqlite.SQLiteDatabase` share same SQLite database file with CBL uses? Or does it create different SQLite database file?

**Comment by snej** [comment]
Even with multiple copies of SQLite in a process, it seems that the file locking problems described only apply if both libraries open _the same database file_. Because the failure scenario seems to be that the two libraries can be unaware of each others' file locks and can end up writing to the file at the same time. I don't see how that applies here, since presumably the other copy of SQLite is being used to work on some files unrelated to Couchbase Lite.

However, I admit it's suspicious that the Mac app exhibiting corruption is also using multiple copies of SQLite...

**Comment by mtrakal** [comment]
Hi @hideki,

we use different SQLite database file for our sqlite, than CBL.

**Comment by pasin** [comment]
> From #966 (comment) that @mtrakal have posted, it seems like some of revision ids could contain non-ascii due to java string formatter issue (that was recently fixed).

> Can this cause an issue with the RevID collator and result to the database corruption?

I have just tested it, and it didn't make the database to corrupt.

**Comment by pasin** [comment]
Some info about the SQLite library, Android SDK will load `libsqlite3.so`, while CBL will load our own built `libsqlite.so` (v. 3.8.10). When CBL JNI is built, it linked with `libsqlite.so`.

**Comment by pasin** [comment]
@snej commented on slack about different SQLite libs:

> It does depend on how the linking is done. But I don’t think the symbols would get mixed up.
For example, I might build dynamic library A that contains libSQLiteA, and separately build library B that contains libSQLiteB. The references to SQLite symbols are resolved at link time for each library, so they don’t interfere with each other when both libraries get loaded in the same process.

**Comment by hideki** [comment]
I tested MultiSQLiteLink scenario. One SQLite database is opened with 	`android.database.sqlite.SQLiteDatabase`, which uses Android provided system SQLite library (3.9.2 for Android N, 3.8.10.2 for Android M, 3.7.11 for Android 4.1.2). Another one is with Couchbase Lite with SQLite storage type, which uses Couchbase provided SQLite library (3.8.x). Test app inserts/reads records in both SQLite database and Couchbase Lite from main UI thread or from background thread with AsyncTask. I tested with Android N, Android M and Android 4.1.2 with real devices. But I can not reproduce the database corruption issue. I also tested the killing the app middle of data insertion. Also I tested scenario of system shutdown caused by out-of-power. Both scenarios can not cause database corruption.

Test app I used: https://github.com/hideki/MultiSQLiteLink

**Comment by hideki** [comment]
Hi @mtrakal,
It is hard to reproduce the problem or to identify the root cause of database corruption issue with only  stack-trace of `SQLiteDatabaseCorruptException`. We need additional information from you. Any additional information is appreciated.
Thanks,

**Comment by hideki** [comment]
Note to self:

@pasin noticed that  [this fix](https://github.com/couchbase/couchbase-lite-java-native/commit/bbfdf5ba922a39b684833a3c82e67a6064cf26d9) may not be enough for some users because corruption occurs in view/query (see: http://crashes.to/s/40365dbe08f). This corruption could be same issue with #846.

ToDo:
- Review com_couchbase_lite_storage_SQLiteJsonCollator.cpp and ICU4C again.

**Comment by pasin** [comment]
@mtrakal, from the link http://crashes.to/s/40365dbe08f to the crashlytics you provide, I noticed that it can show more details if login. I'm wondering if crashlytics collects the logcat log of each crash report or not? Also, is there device locale information shown in the crash report? Anymore info that you can provide can help us investigating this issue.

**Comment by hideki** [comment]
`collateJSON(...)` method of `com_couchbase_lite_storage_SQLiteJsonCollator.cpp` could enter infinite loop if `chars1` or `chars2` are invalid JSON string because do-while end condition is only `depth` variable is `0`. This method does not care input length, `len1` and `len2` values.
```c++
static int collateJSON(void *context, int len1, const void * chars1, int len2, const void * chars2) {
    ...
    int depth = 0;
    ...
    do {
      ...
      ...
    } while (depth > 0); // Keep going as long as we're inside an array or object.
    
    return 0;
}
```
https://github.com/couchbase/couchbase-lite-java-native/blob/501be320fea1396e5922c1ddbde08af4d2e87664/jni/source/com_couchbase_lite_storage_SQLiteJsonCollator.cpp#L411-L507

**Comment by snej** [comment]
`collateJSON` very much requires that its input be valid JSON. It's designed to be as fast as possible, and assumes that validation has occurred before the JSON is added to a row.

In the Cocoa version, the JSON stored in `rows` is always generated by the system `NSJSONSerialization` class — even JSON that comes in from the replicator gets parsed and modified (e.g. removing `_id` and `_rev`) and re-serialized.

**Comment by mtrakal** [comment]
Hi, 

there is a few user traces. I store a data too, but is contains private data, so I will must analyze that and try to find some common signs. For now I have no idea what is it :/.

[fabric4390.zip](https://github.com/couchbase/couchbase-lite-android/files/471999/fabric4390.zip)

**Comment by hideki** [comment]
Hi @mtrakal,

Thank you for sharing additional data.

Note: it seems logs contains totally 15 crash logs from 5 unique devices.

All of crashes occur in following methods. We could focus on View/Query.
```
Caused by com.couchbase.lite.internal.database.sqlite.exception.SQLiteDatabaseCorruptException: database disk image is malformed (code 11)
       at com.couchbase.lite.internal.database.sqlite.SQLiteQueryCursor.nativeMoveToNext(SQLiteQueryCursor.java)
       at com.couchbase.lite.internal.database.sqlite.SQLiteQueryCursor.moveToNext(SQLiteQueryCursor.java:58)
       at com.couchbase.lite.storage.SQLiteStorageEngineBase$SQLiteCursor.moveToNext(SQLiteStorageEngineBase.java:257)
       at com.couchbase.lite.store.SQLiteViewStore.runQuery(SQLiteViewStore.java:1013)
       at com.couchbase.lite.store.SQLiteViewStore.regularQuery(SQLiteViewStore.java:623)
       at com.couchbase.lite.View.query(View.java:358)
       at com.couchbase.lite.Database.queryViewNamed(Database.java:2194)
       at com.couchbase.lite.Query.run(Query.java:446)
```
Thanks!

**Comment by snej** [comment]
Where the crashes occur isn't necessarily related to where the corruption occurs.

**Comment by snej** [comment]
Where is the source code for SQLiteQueryCursor? I couldn't find it in java-core.

**Comment by hideki** [comment]
`SQLiteQueryCursor.moveToNext()`
https://github.com/couchbase/couchbase-lite-java-core/blob/4cc2f7d66a9be58ca774d102497b7a725885e168/vendor/sqlite/src/java/com/couchbase/lite/internal/database/sqlite/SQLiteQueryCursor.java#L58

**Comment by kombadzomba** [comment]
We have received crash from our user related to this issue. We are not able to reproduce it at this moment. Couchbase lite lib is 1.3.0. 

[Details on the crash](http://crashes.to/s/fdd84212cdd).

**Comment by hideki** [comment]
@kombadzomba Thank you for your input.

**Comment by kombadzomba** [comment]
@hideki We got today one more exception from another user. The common with previous one is that the crash occurred when running queries (four queries in different threads) immediately after the database initialization and setting up and triggering push and pull replication. We are already calling some other queries in parallel for some time without problems, but this is the first time we did heavy querying immediately after the database initialization. We hope that postponing these queries will eliminate the code 11 error with our next release.

**Comment by hideki** [comment]
Hi @kombadzomba,
Thank you for your additional inputs. 
- Do you have stacktrace or log messages from logcat when query crashed? I believe not all of your application users face the crash in query. So I assume user who faces the crash in the query might use the locale setting we don't test, or might input special characters like emoji (like other crash issue). Or, from v1.3.0, CBL Java/Android use multiple SQLite connections (maximum is 4), there could be a bug in multiple connection supported codes.
- If possible, can you share 4 map functions? And query codes if app uses query options?

Thanks,
Hideki

**Comment by kombadzomba** [comment]
Hi @hideki,

[Logcat for the first crash](https://gist.github.com/kombadzomba/e38bf561841ed91d91598ad4483c29e3)
[Logcat for the second crash](https://gist.github.com/kombadzomba/b4607431bcd767d4dd8d94b1c29f95ab)
All the documents in the userDb and orgDb are ASCII chars.

[Four warmup threads](https://gist.github.com/kombadzomba/0d7722d71fbd93d65efe3ea59cdb3654)
[Four warmup queries](https://gist.github.com/kombadzomba/d60d8c2d76c828cd1c060e80d3ff6a05)
[view.getData()](https://gist.github.com/kombadzomba/fcd27a0a439aa7413f1f12b4e4e354d4)
[checkIfInitializationDone()](https://gist.github.com/kombadzomba/d1ed45dcd06c5047aba04e39c267e3a9)

Hope this helps.

**Comment by snej** [comment]
Again: I don't think the backtrace of the except is very important, if the db is actually corrupted, because it isn't related to when/where the actual corruption happened. Presumably the db was corrupted the previous time the app ran, and it gets detected the next time the db is opened and read.

It would be very useful to actually get a copy of one of these corrupt database files. Is that possible?

**Comment by mtrakal** [comment]
Hi @snej,

I have some jsons of corupted db, but it looks ok. I dont have local files, just from backend server. Could it help? It contains user data, but I can anonymize it little bit and send it you as a private massage.  Not here. Can it help to you,  when data comes from backend?

**Comment by snej** [comment]
No, sorry. We need the actual SQLite database file with the corruption.

It's not the _data_ in the file that's corrupt. It's the internal structure of the database file itself — the btree and other structures.

**Comment by hideki** [comment]
Hi @kombadzomba,
Thank you for another information.

**Question 1:** Do you use SQLCipher? (SQLite + Encryption)

From `Logcat for the second crash`
```
43	09-15 20:54:42.627 W/Database( 7040): Decrypting database failed
```

**Question 2:** Can you reproduce this by yourself? If yes, can you grab the SQLite file from the device? or Did you collect logcat from the end user?

**Notes:** First crash is caused by `SQLiteDatabaseCorruptException`. It indicates database corruption is not side effect of other crash.
From `Logcat for the first crash`
```
30	09-13 22:45:23.180 E/Event (29395): com.couchbase.lite.internal.database.sqlite.exception.SQLiteDatabaseCorruptException: database disk image is malformed (code 11)
```

Thanks!
Hideki

**Comment by kombadzomba** [comment]
I don't use SQLCipher or any encryption knowingly. We haven't reproduced the issue yet and the logcat I've sent you is from ACRA report, didn't collect anything more than that. Sorry. We'll try to contact the users for more details in order to reproduce.

**Note:** The second user is using 1.3.0 couchbase lite from the beginning so I suppose that migration from previous versions of couchbase lite to 1.3.0 can be eliminated as cause.

**Comment by hideki** [comment]
@kombadzomba I mis-understood our codes. Both SQLite and SQLCipher calls `SQLiteStorageEngineBase.decrypt(...)` no matter database is encrypted. Second log message just indicates database is already corrupted when sending first query to SQLite.

**Comment by hideki** [comment]
Hi @kombadzomba,
In the first crash log from the logcat, can you find the message, "The connection is acquired for a read-only statement."?
message from: https://github.com/couchbase/couchbase-lite-java-core/blob/f25f9a2cbd2ed4f1bb71e07022e6aaa2643267c2/vendor/sqlite/src/java/com/couchbase/lite/internal/database/sqlite/SQLiteConnection.java#L836-L838

**Comment by kombadzomba** [comment]
Unfortunately I have only last 100 lines of logcat from ACRA that I've already shared with you.

**Comment by kombadzomba** [comment]
Hi @hideki 

[Here](https://gist.github.com/kombadzomba/c7fdeb6575d7470855b7d2e2c07e6d23) is the latest logcat of user exception. I thought you may be interested in this line:
`7	09-18 08:28:59.674 W/System.err(26613): com.couchbase.lite.CouchbaseLiteException: Unable to create a storage engine, Status: 401 (HTTP 401 unauthorized)`

**Comment by borrrden** [comment]
I also saw "unauthorized" error during corruption because an unauthorized error looks the same as a corruption error (it means unable to read the database header because it's "corrupted" or "encrypted" and there is no way to tell which)

**Comment by hideki** [comment]
Hi @kombadzomba,
Thank you for new log. New log's problem is caused by `corrupted database`, which is same with your log 2.

**Comment by hideki** [comment]
Hi @kombadzomba,
`1.3.1-40` contains change of [throwing exception if SQLite connection is used for a write operation](https://github.com/couchbase/couchbase-lite-java-core/issues/1446). When you have time, can you try the build,`1.3.1-40` to check if your application throws the `SQLiteException`? Multiple write operations from multiple SQLite connection from multiple threads is suspicious area to cause corruption. 
http://files.couchbase.com/maven2/com/couchbase/lite/couchbase-lite-android/1.3.1-40/
Thanks!

**Comment by kombadzomba** [comment]
@hideki,

The only way to test this build is in production since we are unable to reproduce the corruption. However, I am afraid of doing that with unstable release (potential regression bugs, etc.) but this is special case since it looks that this bug we are talking about is regression bug itself (compared to 1.2.1). We will test if 1.3.0-40 is a candidate for use in production but I am not excited that much about testing this on our already frustrated users to be honest.

**Comment by hideki** [comment]
Hi @kombadzomba,
Sorry for my previous comment, it confused you. `1.3.1-40` is not our official release build. I don't ask you to release your application with 1.3.1-40. I'd like to know if SQLiteException is thrown from your application with normal scenario. Your internal testing is enough.

**Comment by kombadzomba** [comment]
Ok @hideki
Thank you!

**Comment by kombadzomba** [comment]
Hi @hideki, 
There is no `SQLiteException` for now on Nexus 5 with `1.3.1-40`. Will keep testing on other devices tomorrow and let you know if error occurs.

**Comment by hideki** [comment]
@kombadzomba, Thank you for your quick updates!

**Comment by kombadzomba** [comment]
Hi @hideki,

We have published app update with 1.3.1 but the same error ` database disk image is malformed (code 11)` is occurring on new devices. The most devices are SAMSUNG phones.

We really don't know what else to do. We are loosing users because of this bug and it's getting pretty ugly. What are the potential issues of downgrading couchbase lite lib to 1.2.1?

**Comment by hideki** [comment]
Hi @kombadzomba,

I am sorry to hear your problems with v1.3.x.

Can you compile CBL from source codes? Followings are potential problems and solutions.

#### Problem & Solution 1
Since v1.3.0, CBL Android supports multiple SQLite connections (maximum is 4) for Android L (Lollipop) or higher. From @mtrakal 's crash log, his app has faced crash with only Android L and M (N does not use SQLite). To change to only one SQLite connection, please modify `AndroidSQLiteStorageEngine .getWALConnectionPoolSize()` to return always `1`.
https://github.com/couchbase/couchbase-lite-android/blob/master/src/main/java/com/couchbase/lite/android/AndroidSQLiteStorageEngine.java#L58

Note: By setting `1`, any read/write SQLite operations blocks other. 

#### Problem & Solution 2
CBL Android (any versions) creates SQLite connection with `Multi-thread` mode. This mode could cause problem if multiple threads uses same SQLite connection. See https://www.sqlite.org/threadsafe.html . From our code review, This should not be happened. There could be bug that we could not find yet.
 
Opening the database with Serialized mode (see https://www.sqlite.org/threadsafe.html) could solve this issue. Needs to add `SQLITE_OPEN_FULLMUTEX` option to sqliteFlags before calling `sqlite3_open_v2()`?

```sqliteFlags |= SQLITE_OPEN_FULLMUTEX;```

https://github.com/couchbase/couchbase-lite-java-native/blob/688a26047c26c66cdbac68ed33fe2407eb93cd71/jni/source/com_couchbase_lite_internal_database_sqlite_SQLiteConnection.cpp#L90

#### Problem & Solution 3
Since v1.3.0, CBL Android uses Couchbase compiled SQLite native library to support Android N, as Android N does not allow an application to access Android provided native libraries including sqlite(3).so. If This is root cause, solution could be downgrade to CBL v1.2.1. **NOTE:  CBL v1.2.1 can not support Android N.**

Regards,
Hideki

**Comment by hideki** [comment]
Note:

- [Applied Serialized mode for SQLite Connection](https://www.sqlite.org/threadsafe.html)
- [PR](https://github.com/couchbase/couchbase-lite-java-native/pull/53)

**Comment by kombadzomba** [comment]
Hi @hideki, 

It turned out that we had only one crash since we published app with CLB 1.3.1. Even that one was from the user that was using CBL 1.3.0 before and updated app to the new CBL 1.3.1 on the day of the crash. Chances that data corruption occurred before the update are high. That is the reason we have not applied the suggested solutions yet. We will wait for new crashes and keep you updated.

Thanks.

**Comment by snej** [comment]
Crossing our fingers … [strangely, there's no emoji for that!]

**Comment by kombadzomba** [comment]
We have received one crash report with database corruption yesterday from user using only CBL 1.3.1: `database disk image is malformed (code 11)`. User installed fresh application two days before the crash. There was no application update in between.

Device: SM-G930P
Android OS: 6.0.1

Compared to 1.3.0 where we had 1 crash in 24h this is 10 times less.

**Comment by hideki** [comment]
@kombadzomba Thank you for your update.

**Comment by hideki** [comment]
Power failure might cause `database disk image is malformed` issue.
https://github.com/couchbase/couchbase-lite-ios/issues/1482

**Comment by mtrakal** [comment]
Hi @hideki, 

we finally release new version with `cbl 1.3.1-60` on just 5% of all users yesterday. Now we have next 2 users with this error.
http://crashes.to/s/40365dbe08f 
I'll inform you when number will be increased after release on more users.

**Comment by hideki** [comment]
Hi @mtrakal,
Thank you very much for your updates!
CBL Android `1.3.1.-60` uses `Serialized` mode for SQLite threading. https://www.sqlite.org/threadsafe.html It might help to reduce the number of facing corruption issue. Maybe too early to talk.
Thanks again!
Hideki

**Comment by trramsiq** [comment]
Just got this crash today running 1.3.1-60

```
Fatal Exception: com.couchbase.lite.internal.database.sqlite.exception.SQLiteDatabaseCorruptException: database disk image is malformed (code 11)
       at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.nativeExecuteForChangedRowCount(SQLiteConnection.java)
       at com.couchbase.lite.internal.database.sqlite.SQLiteConnection.executeForChangedRowCount(SQLiteConnection.java:617)
       at com.couchbase.lite.internal.database.sqlite.SQLiteSession.executeForChangedRowCount(SQLiteSession.java:725)
       at com.couchbase.lite.internal.database.sqlite.SQLiteStatement.executeUpdateDelete(SQLiteStatement.java:67)
       at com.couchbase.lite.internal.database.sqlite.SQLiteDatabase.delete(SQLiteDatabase.java:1483)
       at com.couchbase.lite.storage.SQLiteStorageEngineBase.delete(SQLiteStorageEngineBase.java:217)
       at com.couchbase.lite.store.SQLiteStore.pruneDocument(SQLiteStore.java:2537)
       at com.couchbase.lite.store.SQLiteStore.add(SQLiteStore.java:1609)
       at com.couchbase.lite.Database.put(Database.java:1976)
       at com.couchbase.lite.Database.putRevision(Database.java:1923)
       at com.couchbase.lite.Database.putRevision(Database.java:1914)
       at com.couchbase.lite.Document.putProperties(Document.java:524)
       at com.couchbase.lite.Document.putProperties(Document.java:350)
       at com.app.DevUtilities.Cache.Cache.store(Cache.java:90)
```

**Comment by hideki** [comment]
Hi @trramsiq,

Thank you for your report. Once SQLite database file is corrupted, it can not be recovered. To avoid many crashes by same user, it requires to delete database or application to create fresh SQLite database file if user can replicate data from the server.

CBL copied SQLiteDatabase classes from android codes. Although original SQLiteDatabase class has logic to delete SQLite file if it faces SQLiteDatabaseCorruptException, but we disabled it. So corrupted SQLite Database file is not automatically deleted.

We are keep eye open to this issue.

Thanks,
Hideki

**Comment by kombadzomba** [comment]
Hi @hideki, 

Can you publish CBL with ability to set parameter saying if database should be deleted or not when SQLiteDatabaseCorruptException occurs? That would do the job in our case since we fetch the data from the server when local database does not exist.

**Comment by hideki** [comment]
@kombadzomba,

Can you capture `SQLiteDatabaseCorruptException` in application level, and delete database, then recover database by replication?

**Comment by kombadzomba** [comment]
I had problems with that solution since exception occurs outside of android Context so I need to do a big code refactoring. I was hoping to skip that :)

**Comment by hideki** [comment]
Hi @mtrakal and @kombadzomba,
According to the report from the @mtrakal 's crash report site, it seems number of crashes which was caused by database corruption is less frequent with using latest build that is build with SQLite Serialized mode. Do you have any update about this issue?

**Comment by kombadzomba** [comment]
Hi @hideki,

We are using couchbase-lite-android: 1.3.1-52 since 11/07/16. 
We had four `SQLiteDatabaseCorruptException - database disk image is malformed (code 11)` since that and the last was on 11/21/16.

We have received yesterday new type of error for the first time:

`com.couchbase.lite.storage.SQLException: Cannot decrypt or access the database
at com.couchbase.lite.storage.SQLiteStorageEngineBase.open(SQLiteStorageEngineBase.java:78)
at com.couchbase.lite.store.SQLiteStore.open(SQLiteStore.java:188)
at com.couchbase.lite.Database.open(Database.java:1251)
at com.couchbase.lite.Manager.openDatabase(Manager.java:335)
at com.couchbase.lite.Manager.getDatabase(Manager.java:299)`

**Comment by hideki** [comment]
Hi @kombadzomba,

Thank you for your update.

With encrypted SQLite database, it throws `SQLiteDatabaseCorruptException` exception if it fails to decrypt.  And we have never encountered `database corruption` internally. So currently `SQLiteDatabaseCorruptException` is tied with database decryption error. We should fix this error message. With standard SQLite library, this error should be caused by actual database corruption.

```java
            database = SQLiteDatabase.openDatabase(path, null,
                    SQLiteDatabase.CREATE_IF_NECESSARY | SQLiteDatabase.ENABLE_WRITE_AHEAD_LOGGING,
                    getWALConnectionPoolSize(),
                    null, new ConnectionListener());
            Log.v(Log.TAG_DATABASE, "%s: Opened Android sqlite db", this);
        } catch(SQLiteDatabaseCorruptException e) {
            hasError = true;
            Log.e(Log.TAG_DATABASE, "Unauthorized to open the SQLite database", e);
            throw new SQLException(SQLException.SQLITE_ENCRYPTION_UNAUTHORIZED,
                    "Cannot decrypt or access the database");
        } catch(com.couchbase.lite.internal.database.SQLException e) {
```

Thanks,

**Comment by mtrakal** [comment]
Hi @hideki,

we have still a few users (but maybe they had corrupted database before upgrade to latest version).
Our released version is (I think) `1.3.1-6x`. Today we will release new version with `1.3.1-89`.
Crashes is less and less.

Few users can still reproduce it between two devices:
- A one device uses some older CBL `1.3.1 stable?` with SQLite, older version app.
- B second one with Nexus 5x Android 7, CBL with forest and version `1.3.1-6x`.

Both devices write to the same object and create conflict. But between version is some different (or we write more/less info to object) and B device try to resolve conflict.
Old A device do nothing, but B device is in infinite loop and create a lot of revisions (18000+) until he has a free space (26 GB on internal storage) and flood storage. 
After crash on malformed disk image (11).
This was reported with new CBL. But must have something to do with old version I think and replication/resolving conflicts.

**Comment by hideki** [comment]
Hi @mtrakal,

- Device B: `malformed disk image (11)` is SQLite specific error message. If you observe this error message with ForestDB enabled, it seems your application keeps using SQLite storage.
- Previously you shared your conflict resolution code, I believe there is possibility conflict resolution code could cause infinite loop. Did you fix it? 

Thanks!

**Comment by mtrakal** [comment]
Hi @hideki,

- Device B: it could be devices after upgrade from Android 6 to Adnroid 7. So, thats right, that they still can use SQLite (my fault, sorry)
- We change it a little bit, but maybe not as well as we should.  I'll try to investigate more.

For now we tracked 44 users with malformed DB (from previous 300+).

**Comment by hideki** [comment]
Current summary of Database corruption issue.

- From 1.3.0, CBL Java enables multiple SQLite connections with Write-Ahead Log (WAL) for Android L (5.0) or newer version. Android KitKat or older is still single connection.
- Some end-users started to face `SQLiteDatabaseCorruptException: database disk image is malformed (code 11)` with v1.3.0 or 1.3.1
- Based on our code review, CBL does not share SQLite Connection between multiple threads. So We used `Multi-Thread` mode (https://www.sqlite.org/threadsafe.html) for v1.3.0 and v1.3.1.
- We switched to `Serialized` mode since `1.3.1-60`, `SQLite can be safely used by multiple threads with no restriction`. 
- After enabled `Serialized` mode, number of causing `SQLiteDatabaseCorruptException` is (significantly?) reduced.
- Some end-users are still facing `SQLiteDatabaseCorruptException`. We assume they installed application with CBL v1.3.0 or 1.3.1, and it make the database file corrupted. Then they upgraded to the latest application with latest CBL which use `Serialized` mode.

Note: There is other root causes of `SQLiteDatabaseCorruptException`. However we could not identify it, and also number of impacted end-users are low.

**Comment by hideki** [comment]
@mtrakal,
FYI: We fixed the code to serialized Map object to JSON. Previously JSON's keys are not ordered correctly. This fix affects to generated Revision ID, and Revision ID value could affect to older of conflicted revision list or winning revision. 
I assume this fix and application's conflict resolution code could cause infinite loop.

**Comment by mtrakal** [comment]
Hi @hideki ,

we still have some issue with mallformed datatabase (malformed (code 11)).

http://crashes.to/s/69021468487

I attached a few logs, look almost same on end lines.

```
Caused by com.couchbase.lite.internal.database.sqlite.exception.SQLiteDatabaseCorruptException: database disk image is malformed (code 11)
       at com.couchbase.lite.internal.database.sqlite.SQLiteQueryCursor.nativeMoveToNext(SQLiteQueryCursor.java)
       at com.couchbase.lite.internal.database.sqlite.SQLiteQueryCursor.moveToNext(SQLiteQueryCursor.java:58)
       at com.couchbase.lite.storage.SQLiteStorageEngineBase$SQLiteCursor.moveToNext(SQLiteStorageEngineBase.java:257)
```

a line before is different:
`at com.couchbase.lite.store.SQLiteStore.getDocument(SQLiteStore.java:779)` and 
`at com.couchbase.lite.store.SQLiteViewStore.runQuery(SQLiteViewStore.java:1030)`

[02.txt](https://github.com/couchbase/couchbase-lite-android/files/613470/02.txt)
[03.txt](https://github.com/couchbase/couchbase-lite-android/files/613468/03.txt)
[04.txt](https://github.com/couchbase/couchbase-lite-android/files/613469/04.txt)

**Comment by hideki** [comment]
Hi @mtrakal,
Crashing 200 times a day is a lot. When capturing the `SQLiteDatabaseCorruptException`, I believe the database is already corrupted. We need to identify the root cause of this. Can you observe any other error messages, exceptions or other indications before `SQLiteDatabaseCorruptException`?

**Comment by mtrakal** [comment]
Hi @hideki,

is there some possible way, how to provide SQLite callback:

```
public class MyDbErrorHandler implements DatabaseErrorHandler {
    @Override
    onCorruption(SQLiteDatabase db) {
       // delete corrupted DB and restart app
    }
}
SQLiteOpenHelper dbHelper = new SQLiteOpenHelper(context, "MyDbName", null, 1,
                                                 new MyDbErrorHandler());
SQLiteDatabase db = dbHelper.getWritableDatabase();

```
as described on this link: http://stackoverflow.com/a/7765336/4024146 ?

We could delete all local data for users, because they can't open app and historical data should be synced to our servers. So after delete DB, restart and relogin data are sync back to phone...

**Comment by hideki** [comment]
Hi @mtrakal,

CBL ported SQLiteDatabase classes from Android, and CBL use it to access SQLite internally. Android version of SQLiteDatabase class is implemented DatabaseErrorHandler, and it deletes the database file if it detects Database corruption. However, when we ported it into CBL, we removed that code as it hides the problem and we considered just deleting the database file is not right solution.

We will consider to provide the API to set DatabaseErrorHandler. 

Thanks,
Hideki

**Comment by hideki** [comment]
Note to Self:
Some of the community users are still facing this issue. However, we don't know the root cause of this issue. I removed the 1.4.0 milestone tag. Keep this ticket in Review column.

**Comment by hideki** [comment]
Hi Guys,

Does your application close the database in [Application.onTerminate()](https://developer.android.com/reference/android/app/Application.html#onTerminate())?

One of our community users has faced DatabaseCorruption issue after upgrade his/her application. I assume that application is terminated without closing the database. It might cause db corruption.
https://github.com/couchbase/couchbase-lite-java-core/issues/1614

Thanks,
Hideki

**Comment by kombadzomba** [comment]
We are not using Application.onTerminate() since we have active Service all the time. We handle replication start/stop in android Service however. Replication is continuous while there is an active application screen.

Instead of Application.onTerminate() we are using Service.onDestroy() where we call Database.compact() 

We still get DatabaseCorruption from users.

**Comment by erpankaj22081982** [comment]
We are also facing this issue very frequently now. along with IP exception in pull replication.
